pcr_profile.go

```diff
@@ -179,28 +144,149 @@ func (p *PCRProtectionProfile) AddProfileOR(profiles ...*PCRProtectionProfile) *
-func (p *PCRProtectionProfile) computePCRValues(tpm *tpm2.TPMContext, values pcrValuesList) (pcrValuesList, error) {
-	if len(values) == 0 {
-		values = append(values, make(tpm2.PCRValues))
+// pcrProtectionProfileIterator provides a mechanism to perform a depth first traversal of instructions in a PCRProtectionProfile.
+type pcrProtectionProfileIterator struct {
+	instrs [][]pcrProtectionProfileInstr
+}
+
+// descendInToProfiles adds instructions from the supplied profiles to the front of the iterator, so that subsequent calls to
+// next will return instructions from each of these profiles in turn.
+func (iter *pcrProtectionProfileIterator) descendInToProfiles(profiles ...*PCRProtectionProfile) {
+	instrs := make([][]pcrProtectionProfileInstr, 0, len(profiles)+len(iter.instrs))
+	for _, p := range profiles {
+		instrs = append(instrs, p.instrs)
 	}
+	instrs = append(instrs, iter.instrs...)
+	iter.instrs = instrs
+}
 
-	for _, instr := range p.instrs {
-		var err error
-		values, err = instr.apply(tpm, values)
-		if err != nil {
-			return nil, err
-		}
+// next returns the next instruction from this iterator. When encountering a branch, a *pcrProtectionProfileAddProfileORInstr will
+// be returned, which indicates the number of alternate branches. Subsequent calls to next will return instructions from each of
+// these sub-branches in turn, with each branch terminating with *pcrProtectionProfileEndProfileInstr. Once all sub-branches have
+// been processed, subsequent calls to next will resume returning instructions from the parent branch.
+func (iter *pcrProtectionProfileIterator) next() pcrProtectionProfileInstr {
+	if len(iter.instrs) == 0 {
+		panic("no more instructions")
+	}
+	if len(iter.instrs[0]) == 0 {
+		iter.instrs = iter.instrs[1:]
+		return &pcrProtectionProfileEndProfileInstr{}
 	}
 
-	return values, nil
+	instr := iter.instrs[0][0]
+	iter.instrs[0] = iter.instrs[0][1:]
+
+	switch i := instr.(type) {
+	case *pcrProtectionProfileAddProfileORInstr:
+		iter.descendInToProfiles(i.profiles...)
+	}
+
+	return instr
+}
+
+// traverseInstructions returns an iterator that performs a depth first traversal through the instructions in this profile.
+func (p *PCRProtectionProfile) traverseInstructions() *pcrProtectionProfileIterator {
+	i := &pcrProtectionProfileIterator{}
+	i.descendInToProfiles(p)
+	return i
+}
+
+// pcrProtectionProfileComputeContext records state used when computing PCR values for a PCRProtectionProfile
+type pcrProtectionProfileComputeContext struct {
+	parent *pcrProtectionProfileComputeContext
+	values pcrValuesList
+}
+
+// handleBranch is called when encountering a branch in a profile, and returns a slice of new *pcrProtectionProfileComputeContext
+// instances (one for each sub-branch). At the end of each sub-branch, finishBranch must be called on the associated
+// *pcrProtectionProfileComputeContext.
+func (c *pcrProtectionProfileComputeContext) handleBranch(n int) (out []*pcrProtectionProfileComputeContext) {
```

Comment:
shouldn't this be called handleBranches or handleBranching ? otherwise it not so clear that is setting state for n OR branches